### DISCUSSION TITLE
Confused by audio ports

### DISCUSSION THREAD

**Discussion by sagantech** [free-audio/clap#169] [Q&A] [answered]
For now I am interested in output audio ports from a host perspective.
 
clap_audio_port_info describes an output configuration or an output bus?
in struct clap_process there is a comment about an index...
 
>  // Audio buffers, they must have the same count as specified
>    // by clap_plugin_audio_ports->get_count().
>    // The index maps to clap_plugin_audio_ports->get_info().
>    const clap_audio_buffer_t *audio_inputs;
<snip>

What index? 

If a plugin, for example, supports 8 audio output channels, say 3 stereo and 2 mono, how would that be configured?
How does 'process' render to multiple buses? Is it all at once or it needs to make multiple calls?

Any help greatly appreciated.

**Comment by baconpaul** [comment, marked as answer]
`clap_plugin_audio_ports` has a `count` method which says how many ports you have. Each port has a bus configuration which is described by the `clap_audio_port_info` which is return by `_audio_ports::get` so in your example the plugin would return

`_audio_ports::count` -> 5

`_audio_ports::get(0..2)` -> return an info with `info::channel_count == 2`

`_audio_ports::get(3..4)` -> return an info with `info::channel_count == 1`

`process` has a member `*audio_inputs` which is of type `clap_audio_buffer`. In this case a properly configured host would meet constraints like

`process->audio_inputs[0]->channel_count = 2`

and

`process->audio_inputs[3]->channel_count = 1`

similarly the `**data32` member of `audio_inputs[0]` would have size `[2][frames]` and of `[3]` would be `[1][frames]`

Does that help?

**Reply by sagantech** [reply]
Thanks for the reply. I will look at this and mark it answered when I get it working. It does definitely help. 
BTW I downloaded Surge XT and the new browser is **beautiful**.

**Reply by baconpaul** [reply]
Thank you! Did you try the MSEG and tuning editor (and mod list now I think of it)? We are really proud of those also.

**Reply by baconpaul** [reply]
but its come a long way since you fixed the vstgui windows checkmark implementation in 1.6 beta 7 or whenever! (We also almost entirely eliminated allocations on the audio thread, which was another of the hairy issues you raised back then; but that was hairy and theres a few small edges we can't get 100% still).

**Reply by sagantech** [reply]
Glad to see that issue is still a priority. I was under the impression the team thought that it was a non-issue. As I am sure you know, standard memory allocations on an audio thread will eventually result in a priority inversion.

I realize this is the wrong forum but I am also having a problem with Surge XT window resizing. I should say that my (Windows x64) clap implementation is quite new and I have only tried two plugs thus far; Full Bucket theScrooo and SurgeXT. Both are working fine. Resizing with theScroo works fine. Resizing SurgeXT to a smaller size works fine. When resizing to a larger size, the window will not refresh the newly uncovered portion unless I hover over, manually change the frame or move it offscreen and then back on. Any ideas?

**Reply by sagantech** [reply]
So to be sure I got this right. The CLAP version of Surge XT has 3 separate stereo outs for a total of 6 audio outs. Correct?

**Reply by baconpaul** [reply]
Yes

**Reply by baconpaul** [reply]
The resizing tbing sounds like your resize isn’t forcing a refresh which should happen via juce when sersize is called. Is your drag gesture calling those apis in the right order? But happy to open a separate issue

**Reply by baconpaul** [reply]
Oh and the memory thing was a priority - esepcialky the allocators which didn’t free - but it was really annoyingly hard to fix the main culprit. But I think we got it around 1.9 or so

**Reply by sagantech** [reply]
Thanks for the quick responses. They are greatly appreciated.

> The resizing tbing sounds like your resize isn’t forcing a refresh which should happen via juce when sersize is called. Is your drag gesture calling those apis in the right order? 

I am not resizing. I am simply right-clicking the Surge gui and selecting a larger size. The rest of your sentence is difficult for me to parse. I am not using Juce (I know that you are) but what is "sersize" and what is the "drag gesture" apis in the right order? Please can we open a separate issue?

**Reply by sagantech** [reply]
I just tried U-HE MFM2 and it resizes fine also so it may be a surge issue.

**Reply by baconpaul** [reply]
Ahh its probably the juce adapter. So you choose "zoom to 140" and you get a misdraw? I wonder which other zoomable juce plugins you can try. (FullBucket is iplug; U-He is hand rolled; Surge and others are JUCE).

Is there any way i can try your host? we work fine in the other hosts we've tried but these zoom protocols are tricky and maybe we are doing something invalid-but-working which you've found (or maybe we are doing something valid-but-not-working which you need to fix; it's been a bit of both as we've trundled along the clap path).

And no problem on quick responses! Happy to help! But I have some travel next couple of days so responsiveness will drop some.

**Reply by sagantech** [reply]
Yes I tried Surge with the host supplied here and it resizes fine. I would be glad to send you a copy although you should know I have only been working on CLAP for about a week so there are bound to be issues even though all the basics seems to be working quite well. CLAP is a welcome improvement in most areas. Can you send me an email (info_at_sagantech.biz) ? Do so and I will set you up.
